README.md

```diff
@@ -1 +1,64 @@
-# mailgun2
+to build run: `cd service ` then `go build`
+
+to start the service you can run: `./service` the service defaults to port 1337
+
+you can override the port it listens to by passing the port flag like this `./service --port=1338`
+
+
+i created a local docker file for postgres and ran it on a custom port running the following command:
+`docker run -d --name mailgun-postgres -e POSTGRES_PASSWORD=Pass2020! -v ${HOME}/postgres-data/:/var/lib/postgresql/data -p 6001:5432 postgres`
+
+then rerunning that docker with this command:
+`docker start <uuid of mailgun-postgres>`
+
+
+Design 
+The code is a simple implementation that can be scaled horizontally
+
+The major issue with this system is the massive amount of writes being performed. 
+Generally systems are dominated by reads, facebook/amazon etc. 
+However this system is dominated by writes
+
+Since the writes are updating/creating a domain with 1 value we can reduce the writes by adding a simple
+messaging queue that could aggregate writes over a certain time interval and then flush those updates. 
+![Alt text](pictures/code_diagram.png?raw=true "Code Diagram")
+
+We can expand this design further by adding a loadbalancer between the client, and the service
+
+![Alt text](pictures/scale1.png?raw=true "Scale Option 1")
+
+We can extend this design also by adding read shards for the database to reduce the load on the
+master write DB
+
+![Alt text](pictures/scale2.png?raw=true "Scale Option 2")
+
+We can still improve upon this design by taking advantage of the domain rules. We can split the datapath
+into starting letters One database could hold all the domains starting with an A then another with B ...until we reach Z
```

Comment:
How could you handle database "P" going down? What potential pitfalls does having this many databases pose?

Reply:
Hi Michael,
I appreciate your feedback.
for the scenario where database P goes down, we could have a read-only replica of the database to fall back to. But since this system seems more write-intensive we may consider having a write clone as a backup. Once failures are detected we would switch over to the clone and then when the DB comes back up with sync those together. There would be some data-consistency issues. Since we want to keep our Integrity our Redundancy went up and our consistency has gone down in this case. 

The pitfalls of having this many databases are the number of failures that will occur if DB1 has an availability of 99.99% this means one database is down for roughly 53 mins a year. multiply that by the number of letters 26* 53 mins = roughly 22.9 hours a year. You most likely would be in violation of your SLA at that point so a hit to data consistency to increase redundancy to improve your availability would be your only choice